### DISCUSSION TITLE
Typeshed versioning using git submodules

### DISCUSSION THREAD

**Discussion by kumaraditya303** [microsoft/pyright#1325] [Ideas]
Currently Typeshed version in pyright is updated manually, it would be easier and more maintainable to do this using git [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules)

https://github.com/kumaraditya303/pyright/tree/submodules/packages/pyright-internal @erictraut

**Comment by erictraut** [comment]
Thanks for your suggestion for how to make it easier for me to maintain the typeshed stubs in pyright, but the update needs to remain manual for multiple reasons:
1. Pyright occasionally introduces new features (e.g. draft PEP support) that require stub modifications that are not yet ready to be accepted into typeshed. A current example of this is the support for `TypeGuard` in `typing_extensions.pyi` ([PEP 647](https://www.python.org/dev/peps/pep-0647/)). This requires some manual merging.
2. Pyright doesn't support Python 2.x, so we don't want to include the directories in typeshed that are specific to 2.x.
3. Typeshed stubs sometimes introduce breaking regressions that we need to detect and report before accepting the latest versions. I run smoke tests on a large internal code base before I accept any updates from typeshed, and this can't be automated in the public github repo.
4. It's important for me to know about any changes in the core typeshed stubs typing.pyi and builtins.pyi because there are some assumptions built into the type checker about some of these declarations. Manually comparing and merging is the best way for me to examine all such changes.
